### ISSUE TITLE
magic dimension names in methods (Time in Tracks layer)

### ISSUE THREAD

**Issue by d-v-b** [napari/napari#1776]
The tracks layer (which looks really cool!) has some references to "time" in documentation and some methods, e.g. https://github.com/napari/napari/blob/master/napari/layers/tracks/tracks.py#L312 Would it make sense for napari to adopt a policy of not doing this? I think it's always a mistake for a domain-agnostic tool to start carrying around assumptions about what array axes mean.

**Comment by sofroniewn** [comment]
> Would it make sense for napari to adopt a policy of not doing this? I think it's always a mistake for a domain-agnostic tool to start carrying around assumptions about what array axes mean.

Yeah, we might want to change that one!

I've been thinking that a lot of the machinery in the tracks layer could be used for things like rendering branching neurons - see discussion in #693 a while ago, in which case definitely dropping that makes sense.

Are you using tracks for something beyond the traditional tracking examples we've seen so far @d-v-b?

@jni @quantumjot @kevinyamauchi - we could try and change this before 0.4.0, or wait till later, curious what others think.

**Comment by d-v-b** [comment]
I don't have any particular use for the tracks layer at the moment but I can think of a lot of uses for a graph layer in EM connectomics, and there's no time axis there :)

**Comment by jni** [comment]
@d-v-b yeah, my goal was always to extract some of the logic from Tracks into a more generic Graph layer. I think there's a lot of things that need to be changed though. For example, are "tracklets" (non-branching paths) a useful concept to keep in a generic Graph layer? Or do they just complicate the logic?

I don't think we should hold off on releasing because of this — Tracks is already insanely useful and we should get it out there for people to use. Originally Image, Volume, and Pyramid were all different layers before @sofroniewn joined them (still one of my favourite efforts in the history of napari @sofroniewn! 😅), and I think we can follow a similar path here.

As a side note @AbigailMcGovern ran into the limitation that Tracks are not actually nD, which violates one of our founding principles. 😄 Again, I think we press on and fix this in the 0.4 series.

**Comment by sofroniewn** [comment]
Ok, moving this to `0.4` instead of `0.4.0` as agreed no need to hold up release, but we can continue to improve with time

**Comment by quantumjot** [comment]
Interesting discussion! A more general `Graph` layer would be really cool. Couple of extra thoughts:
- The concept of time in `Tracks` could be replaced with `index` to make it application agnostic but retain the ordering of the data.
- For other applications, like rendering branching neurons, you could modify the shader to use all of the non displayed dimensions and then do the nD slicing directly (within a user defined window, similar to `tail_length`) in the shader. At the moment, this is only done with the existing `time` dimension.  I can see quite a few advantages to this approach, and it would make the layer much more general than just `Tracks`. In this model, no assumptions are made about what the axes mean.

**Comment by sofroniewn** [comment]
> Interesting discussion! A more general Graph layer would be really cool. Couple of extra thoughts:

I really like the idea that's been proposed. I might start by making a new layer `Graph` that is strictly more general, and then we could alias `Tracks` for that at some point, possibly with some special casing, but that would let us experiment with the more general while maintaining the current behavior.

@adamltyson might be interested in this too. It might also be a way to resolve #1334 on visualizing a 3D path in 2D
